server/auth/types/basic/routes.ts

```diff
@@ -0,0 +1,271 @@
+/*
+ *   Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License").
+ *   You may not use this file except in compliance with the License.
+ *   A copy of the License is located at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   or in the "license" file accompanying this file. This file is distributed
+ *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *   express or implied. See the License for the specific language governing
+ *   permissions and limitations under the License.
+ */
+
+import { IRouter, SessionStorageFactory, KibanaRequest } from "../../../../../../src/core/server";
+import { SecuritySessionCookie } from "../../../session/security_cookie";
+import { SecurityPluginConfigType } from "../../..";
+import { AuthConfig } from "./basic_auth";
+import { filterAuthHeaders } from "../../../utils/filter_auth_headers";
+import { User } from "../../user";
+import { SecurityClient } from "../../../backend/opendistro_security_client";
+import { schema } from '@kbn/config-schema';
+import { CoreSetup } from "../../../../../../src/core/server";
+
+export class BasicAuthRoutes {
+  constructor(private readonly router: IRouter,
+    private readonly config: SecurityPluginConfigType,
+    private readonly sessionStorageFactory: SessionStorageFactory<SecuritySessionCookie>,
+    private readonly securityClient: SecurityClient,
+    private readonly authConfig: AuthConfig,
+    private readonly coreSetup: CoreSetup) {
+  }
+
+  public async setupRoutes() {
+    const PREFIX = '';
+
+    // if the user can be authenticated using auth headers, redirect to the next url, otherwise, render login page 
+    this.router.get(
+      {
+        path: `${PREFIX}/login`,
+        validate: false,
+        options: {
+          authRequired: false,
+        }
+      },
+      async (context, request, response) => {
+        try {
+          const alternativeHeaders = this.config.basicauth.alternative_login.headers;
+          if (alternativeHeaders && alternativeHeaders.length) {
+            let requestHeaders = Object.keys(request.headers).map(header => header.toLowerCase());
```

Comment:
np: Could just create object instead of an array to simplify finding in next line.

Reply:
thanks for the tip. we decided to move the login page rendering to browser side, so I removed this route handler. Will try to adopt this approach in future cases